cf-check/replicate_lmdb.c

```diff
@@ -0,0 +1,270 @@
+{
+    Log(LOG_LEVEL_ERR, "%s err(%d): %s\n", db_file, rc, mdb_strerror(rc));
+}
+
+static void HandleSrcLMDBCorruption(MDB_env *env, const char *msg)
+{
+    LMDBReplicationInfo *info = mdb_env_get_userctx(env);
+    Log(LOG_LEVEL_ERR, "Corruption in the source DB '%s' detected! %s",
+        info->s_file, msg);
+    mdb_env_set_assert(env, (MDB_assert_func*) NULL);
+    if (info->d_txn != NULL)
+    {
+        mdb_txn_commit(info->d_txn);
+    }
+    if (info->s_txn != NULL)
+    {
+        mdb_txn_abort(info->s_txn);
+    }
+
+    /* remove files that LMDB creates behind our back */
+    char *garbage_file = StringFormat("%s-lock", info->d_file);
+    unlink(garbage_file);
+    free(garbage_file);
+
+    exit(CF_CHECK_LMDB_CORRUPT_PAGE);
+}
+
+static void HandleDstLMDBCorruption(MDB_env *env, const char *msg)
+{
+    LMDBReplicationInfo *info = mdb_env_get_userctx(env);
+    Log(LOG_LEVEL_ERR, "Corruption in the new DB '%s' detected! %s",
+        info->d_file, msg);
+    mdb_env_set_assert(env, (MDB_assert_func*) NULL);
+    if (info->d_txn != NULL)
+    {
+        mdb_txn_abort(info->d_txn);
+    }
+    if (info->s_txn != NULL)
+    {
+        mdb_txn_abort(info->s_txn);
+    }
+
+    /* remove files that LMDB creates behind our back */
+    char *garbage_file = StringFormat("%s-lock", info->d_file);
+    unlink(garbage_file);
+    free(garbage_file);
+
+    /* This should actually never happen -- how can there be a corruption in a
+     * freshly created LMDB where we are just inserting data? Hence,
+     * UNKNOWN error. */
+    exit(CF_CHECK_UNKNOWN);
+}
+
+/**
+ * Replicate an LMDB file by reading it's entries and writing them into a new LMDB file.
+ *
+ * @return  CFCheckCode code
+ * @WARNING This function can exit the calling process in case of an LMDB
+ *          operation failure.
+ */
+int replicate_lmdb(const char *s_file, const char *d_file)
+{
+    MDB_env *s_env = NULL;
+    MDB_txn *s_txn = NULL;
+    MDB_dbi s_dbi;
+    bool close_s_dbi = false;
+    MDB_cursor *s_cursor = NULL;
+    MDB_env *d_env = NULL;
+    MDB_txn *d_txn = NULL;
+    MDB_dbi d_dbi;
+    bool close_d_dbi = false;
+    MDB_cursor *d_cursor = NULL;
+    LMDBReplicationInfo info = { s_file, d_file, NULL, NULL };
+
+    int ret = 0;
+    int rc;
+
+    Log(LOG_LEVEL_INFO, "Replicating '%s' into '%s'", s_file, d_file);
+
+    rc = mdb_env_create(&s_env);
+    if (rc != 0)
+    {
+        ret = rc;
+        report_mdb_error(s_file, rc);
+        goto cleanup;
+    }
+    mdb_env_set_userctx(s_env, &info);
+    mdb_env_set_assert(s_env, (MDB_assert_func*) HandleSrcLMDBCorruption);
+
+    /* MDB_NOTLS -- Don't use Thread-Local Storage
+     * Should allow us to use multiple transactions from the same thread (only
+     * one read-write). */
+    rc = mdb_env_open(s_env, s_file, MDB_NOSUBDIR | MDB_RDONLY | MDB_NOTLS, 0600);
+    if (rc != 0)
+    {
+        ret = rc;
+        report_mdb_error(s_file, rc);
```

Comment:
would it be helpful in diagnosing problems to report which of these many steps are failing? Maybe as a parameter to the report_mdb_error() function?

Reply:
Good idea! I'll add it.